### PR TITLE
STRAT-17 - State of Charge Investigation Preliminary

### PR DESCRIPTION
**Pull request by mbA2D**
SoC Algorithm based on the energy left in the pack. Not sure how we want to remove energy from the virtual pack, so made a basic function to do so.

Also a basic model for cell efficiency and associated ohmic power losses

Need to talk with Clarke about how this will fit with everything.

Note - not tested or anything, just programmed it in Notepad++

### CODE REVIEW

File: soc/PackEfficiency.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+	Connector resistance ~= 1mOhm
+	Total = ~55mOhm
+	
+	assume this is constant for now, implement changes for temp, SoC variations later
+	power loss in a pack is due primarily to ohmic heating
+	'''
+	
+	
+	def __init__(self):
+		_pack_resistance = 0.055
```
Comment (clarkevandenhoven) [inline soc/PackEfficiency.py]: This should be declared as self._pack_resistance, also any reason for _pack_resistance instead of pack_resistance?

**Comment by clarkevandenhoven** [comment]
High Level comment, we should try to include at least basic unit testing. I would rather not merge in untested code into the codebase. See https://doc.pytest.org for a great testing library

File: soc/PackEfficiency.py (hunk 1 above)
Comment (mbA2D) [reply soc/PackEfficiency.py]: Copying Karl's style from his proposed BMS SoC algorithm, that's it.
I will try to do some testing before Saturday.

File: soc/SoCEstimation.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+	3.635V nominal
+	'''
+	def __init__(self):
+		self._energy_available = 0
+		self._pack_energy = 36*36*3.45*3.635
+		self._SoC = self._energy_available / self._pack_energy
+		
+	def set_SoC(self, soc):
+		self._SoC = soc
+		self._energy_available = soc * _pack_energy
```
Comment (clarkevandenhoven) [inline soc/SoCEstimation.py]: needs to be self._pack_energy

File: soc/PackEfficiency.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+	
+	#if we want say 1000W at the motors, we will need to draw more power (1050W) from the cells
+	def draw_power(self, power_outside_pack_W):
+		pack_ocv = get_pack_ocv() #from SoC-OCV curve or telemetry, assume cells are balanced
```
Comment (clarkevandenhoven) [inline soc/PackEfficiency.py]: Need get_pack_ocv() to be mocked for testing

**Pull request merged**